File: 프로그래머스/lv2/오하늘/뒤에 있는 큰 수 찾기.py
# 가장 가까이 큰 수가 핵심
# 스택으로 접근해보았으나 시간 초과..
def solution(numbers):
    answer = [-1]

    stack = [numbers[len(numbers)-1]]
    for i in range(len(numbers)-2,-1,-1):

        while stack and numbers[i] >= stack[-1]:
            stack.pop() # 작은 경우만 pop

        if stack:
            answer.insert(0,stack[-1])
        else:
            answer.insert(0,-1)

        stack.append(numbers[i])
            
    return answer

print(solution([9, 1, 5, 3, 6, 2]))

File: 프로그래머스/lv2/이선영/뒤에 있는 큰 수 찾기.java
//https://school.programmers.co.kr/learn/courses/30/lessons/154539

import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] arr = new int[numbers.length];
        Stack<int[]> stack = new Stack<>();
        
        for(int i=0; i<numbers.length; i++){
            int num = numbers[i];
            
            while(!stack.isEmpty() && stack.peek()[1] < num){
                int[] now = stack.pop();
                arr[now[0]] = num;
            }
            
            stack.push(new int[]{i, num});
        }
        
        while(!stack.isEmpty()){
            int[] now = stack.pop();
            arr[now[0]] = -1;
        }
        
        return arr;
    }
}
